File: ProblemSolving/450DSA/Python/src/binarytree/LevelOrderTraversal.py
"""
@author Anirudh Sharma

Given a binary tree, find its level order traversal.
Level order traversal of a tree is breadth-first traversal for the tree.

Constraints:
1 <= Number of nodes<= 10^4
1 <= Data of a node <= 10^4
"""


def levelOrderTraversal(root):
    # Special case
    if root is None:
        return None
    # List to store the result
    result = []
    # Queue to store nodes of the tree
    nodes = []
    # Add root node to the queue
    nodes.append(root)
    # Loop until the queue is empty
    while nodes:
        # Get the current node
        current = nodes.pop(0)
        # Add this node to the result
        result.append(current.data)
        # Check if the left child exists
        if current.left:
            nodes.append(current.left)
        # Check if the right child exists
        if current.right:
            nodes.append(current.right)
    return result


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


if __name__ == "__main__":
    root = Node(1)
    root.left = Node(3)
    root.right = Node(2)
    print(levelOrderTraversal(root))

    root = Node(10)
    root.left = Node(20)
    root.right = Node(30)
    root.left.left = Node(40)
    root.left.right = Node(60)
    print(levelOrderTraversal(root))

File: ProblemSolving/450DSA/JavaScript/src/binarytree/LevelOrderTraversal.js
/**
 * @author Anirudh Sharma
 * 
 * Given a binary tree, find its level order traversal.
 * Level order traversal of a tree is breadth-first traversal for the tree.
 * 
 * Constraints:
 * 1 <= Number of nodes<= 10^4
 * 1 <= Data of a node <= 10^4
 */
const levelOrderTraversal = (root) => {
    // Special case
    if (root === undefined) {
        return undefined;
    }
    // Array to store result
    const result = [];
    // Queue to store nodes of the tree
    const nodes = [];
    // Add root to the queue
    nodes.push(root);
    // Loop until the queue is empty
    while (nodes.length > 0) {
        // Get the current node from the head
        let current = nodes.shift();
        // Add current node to the result
        result.push(current.data);
        // Check if the left child exists
        if (current.left !== null) {
            nodes.push(current.left);
        }
        // Check if the right child exists
        if (current.right !== null) {
            nodes.push(current.right);
        }
    }
    return result;
};

function Node(data, left, right) {
    this.data = (data === undefined ? 0 : data);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

const main = () => {
    let root = new Node(1);
    root.left = new Node(3);
    root.right = new Node(2);
    console.log(levelOrderTraversal(root));

    root = new Node(10);
    root.left = new Node(20);
    root.right = new Node(30);
    root.left.left = new Node(40);
    root.left.right = new Node(60);
    console.log(levelOrderTraversal(root));
};

main();